binary_search_tree/tree.py

```diff
@@ -6,45 +6,124 @@ def __init__(self, key, val = None):
         self.key = key
         self.value = val
         self.left = None
-        self.right = None
-        
+        self.right = None  
 
 
 class Tree:
     def __init__(self):
         self.root = None
 
-    # Time Complexity: 
-    # Space Complexity: 
+    def add_helper(self, current_node, key, value):
+        if current_node == None:
+            return TreeNode(key, value)
+
+        if key <= current_node.key:
+            current_node.left = self.add_helper(current_node.left, key, value)
+        else:
+            current_node.right = self.add_helper(current_node.right, key, value)
+
+        return current_node
+
+    # Time Complexity: O(log n) - if balanced
+    # Space Complexity: O(n) eventually O(log n) - recursive solution, "addresses are removed from stack when returning"
     def add(self, key, value = None):
-        pass
+        if self.root == None:
+            self.root = TreeNode(key, value)
+        else:
+            self.add_helper(self.root, key, value)
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(log n)
+    # Space Complexity: O(n) eventually O(log n) - recursive solution, "addresses are removed from stack when returning"
     def find(self, key):
-        pass
+        if self.root == None:
+            return None
+        
+        current = self.root
+        while current != None:
+            if current.key == key:
+                return current.value
+            elif current.key < key:
+                current = current.right
+            else:
+                current = current.left
+        
+        return None
+
+    # Inorder helper function  
+    def inorder_helper(self, current, traversal_list):
+        if current != None:
+            self.inorder_helper(current.left, traversal_list)
+            traversal_list.append({'key': current.key, 'value': current.value})
+            self.inorder_helper(current.right, traversal_list)
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # NOTE I've been reading different arguments for recursive tree traversals online
+    # and am now unsure of the most "correct" answer
```

Comment:
There are several ways to do it.  Mine gets a list of node values as you traverse, so it can be tested more easily.